homeassistant/components/sonos/__init__.py

```diff
@@ -335,19 +335,30 @@ def _add_speaker(
         except (OSError, SoCoException, Timeout) as ex:
             _LOGGER.warning("Failed to add SonosSpeaker using %s: %s", soco, ex)
 
+    async def _async_gethostbyname(self, host: str) -> str:
+        """Enable patching of this function for unit testing."""
+        return await self.hass.async_add_executor_job(socket.gethostbyname, host)
+
     async def async_poll_manual_hosts(
         self, now: datetime.datetime | None = None
     ) -> None:
         """Add and maintain Sonos devices from a manual configuration."""
-        for host in self.hosts:
-            ip_addr = await self.hass.async_add_executor_job(socket.gethostbyname, host)
+
+        # Loop through each configured host and verify that Soco attributes are available for it.
+        for host in self.hosts.copy():
```

Comment:
This is a change in behavior. The original intention here was to obtain the visible zones from _any_ host and then break. 

That being said, I'm not sure that attempting to detect if any hosts have become unavailable in this manner will be reliable with this method. The call is made to `sync_get_visible_zones` here: https://github.com/home-assistant/core/blob/95d16c98293972e3474f27182ac1bd3e50deab7a/homeassistant/components/sonos/helpers.py#L122-L126

The first two attributes return cached values (no I/O attempted) and the last is aggressively cached from the ZoneGroupState, which is shared across all speakers in the system. This means it's very likely that only the first speaker in a loop will actually be tested (and possibly not even that one!).

If you want to actively ensure all speakers are still available, you'd need to make a non-caching call to each device, such as with this method: https://github.com/home-assistant/core/blob/95d16c98293972e3474f27182ac1bd3e50deab7a/homeassistant/components/sonos/speaker.py#L573-L578

Reply:
On startup with this scenario, the function was running into trouble.  Assume 3 hosts  listed A,B,C; A+C are on and B is off.  The first call to get visible zones for A succeeds and returns A+C, this causes the top loop to exit.

In the bottom loop, it will loop through A,B,C; when it hits B it determines it is not a known speaker and attempts to send a discovery message, which attempt to access soco.uid which is not known or cached and hence performs a blocking network call which generates an exception related to blocking the event loop which was unhandled and this would then cause setup to fail.

What I think we should try to ensure in the top loop is that we have the info cached. Such that we can exclude hosts in the error list to avoid a) logging lots of messages (prior pull request) and b) avoid event loop blocking calls in the lower loop.

When A,B,C are on, yes we will loop three times at the top, but as you state, all that information comes from cache so it's not a large perf hit.

Certainly the comment could be better.